tests/unit/sagemaker/serve/builder/test_model_builder.py

```diff
@@ -2786,3 +2827,158 @@ def test_optimize_for_hf_without_custom_s3_path(
+        - Optimizations that use Quantization must be provided with Compilation for GPU instances.
+        - Optimizations that use Quantization:smoothquant are not supported for GPU instances.
+        """
+        self.assertRaisesRegex(
+            ValueError,
+            textwrap.dedent(expected_compilation_quantization_error_message),
+            lambda: _validate_optimization_configuration(
+                instance_type="ml.g5.24xlarge",
+                quantization_config={
+                    "OverrideEnvironment": {"OPTION_QUANTIZE": "smoothquant"},
+                },
+                sharding_config=None,
+                speculative_decoding_config=None,
+                compilation_config=None,
+            ),
+        )
+
+        expected_quantization_error_message = """
+        Optimization cannot be performed for the following reasons:
+        - Optimizations that use Quantization:test are not supported for GPU instances.
+        """
+        self.assertRaisesRegex(
+            ValueError,
+            textwrap.dedent(expected_quantization_error_message),
+            lambda: _validate_optimization_configuration(
+                instance_type="ml.g5.24xlarge",
+                quantization_config={
+                    "OverrideEnvironment": {"OPTION_QUANTIZE": "test"},
+                },
+                sharding_config=None,
+                speculative_decoding_config=None,
+                compilation_config=None,
+            ),
+        )
+
+    def test_neuron_configurations_throw_errors_for_rule_set(self):
+        self.assertRaisesRegex(
+            ValueError,
+            "Optimizations that use Speculative Decoding are not supported on Neuron instances.",
+            lambda: _validate_optimization_configuration(
+                instance_type="ml.inf2.xlarge",
+                quantization_config=None,
+                speculative_decoding_config={"key": "value"},
+                compilation_config=None,
+                sharding_config=None,
+            ),
+        )
+
+        self.assertRaisesRegex(
+            ValueError,
+            "Optimizations that use Sharding are not supported on Neuron instances.",
+            lambda: _validate_optimization_configuration(
+                instance_type="ml.inf2.xlarge",
+                quantization_config=None,
+                speculative_decoding_config=None,
+                compilation_config=None,
+                sharding_config={"key": "value"},
+            ),
+        )
+
+    def test_trt_configurations_rule_set(self):
+        # Can be compiled with quantization
+        _validate_optimization_configuration(
+            instance_type="ml.g5.24xlarge",
+            quantization_config={
+                "OverrideEnvironment": {"OPTION_QUANTIZE": "smoothquant"},
+            },
+            sharding_config=None,
+            speculative_decoding_config=None,
+            compilation_config={"key": "value"},
```

Comment:
Compilation throught TRT does not have any required variables, so it should work with an empty dict. Same for Neuron.

Reply:
verified